### ISSUE TITLE
The game crashes for people using Chinese input method

### ISSUE THREAD

**Issue by murpii** [ddnet/ddnet#6668]
Reported by kyokaiwei on Discord:
>ws — Today at 20:40
murpi this time update
the game will crash for people using Chinese input method
me and my friends game crashes

Crash Dump:
[CrashRTP.zip](https://github.com/ddnet/ddnet/files/11578962/CrashRTP.zip)

**Comment by Jupeyy** [comment]
```
0x000000014000708d: CCommandProcessorFragment_OpenGL::RunCommand(CCommandBuffer::SCommand const*) at src/engine/client/backend/opengl/backend_opengl.cpp:1131
0x0000000140017b66: CCommandProcessor_SDL_GL::RunBuffer(CCommandBuffer*) at src/engine/client/backend_sdl.cpp:336
0x0000000140014063: CGraphicsBackend_Threaded::ThreadFunc(void*) at src/engine/client/backend_sdl.cpp:65
0x00000001401b914a: CWindowsComLifecycle::~CWindowsComLifecycle() at src/base/system.cpp:4386
```

Mhh I can just guess, but it kinda sounds like it crashes because the focus is somehow changed, because of a faulty driver. Who knows what even the headless IME triggers under the hood.

If you have contact to these players u could ask them for their GPUs and drivers. And afterwards if an driver update + Windows update fixes it

**Comment by murpii** [comment]
I'll ask.

**Comment by murpii** [comment]
kyokaiwei:
![image](https://github.com/ddnet/ddnet/assets/60477660/7bcebf61-c5d9-4d7c-8bad-d2848ca4a2bf)

someone else (not specified who):
![image](https://github.com/ddnet/ddnet/assets/60477660/555fee68-14df-4f61-addb-213a4d171657)
gt1030 31.0.15.1694

**Comment by Jupeyy** [comment]
The first one could defs update his drivers and also try vulkan, if he doesn't (or opengl analogically)

The second is pretty old, but maybe also supports the other backends.

Generally both are old, first is outdated, second is simply a very old GPU.

**Comment by murpii** [comment]
Does having a driver from the previous year truly qualify as "pretty old"? That's abit concerning.

**Comment by Jupeyy** [comment]
Tbh the age of the driver doesn't matter. If it's buggy it's buggy. If they indeed used OpenGL you also have to consider that almost no app or game uses it, especially on Windows. So there is always a potential for it

**Comment by Robyt3** [comment]
1. Does it also crash when using native IME (`inp_ime_native_ui 1`)?
2. Which keyboard are they using for Chinese input in Windows? For me the language settings look like this: 
![chinese-language-settings-windows](https://github.com/ddnet/ddnet/assets/23437060/1ee124d5-17f4-4c73-9e15-938c63208465)
3. Does 17.0 still crash when copying the `SDL.dll` from 16.9 into the 17.0 folder?

**Comment by murpii** [comment]
1. Still crashing
2. 
![image](https://github.com/ddnet/ddnet/assets/60477660/37b4cfd1-ff4a-4440-a319-2f9d16d60ec5)
3. No crash

**Comment by Robyt3** [comment]
So it seems like it's an issue with SDL and we should report it there.

**Comment by Jupeyy** [comment]
Is there any difference theoretically between old SDL2.dll and using the current one with inp_ime_native_ui 1?
Or should:tm: it behave the same?

**Comment by Robyt3** [comment]
Old SDL didn't have the hint that's set when using `inp_ime_native_ui 1`. I'd expect that it doesn't behave the same.

**Comment by Jupeyy** [comment]
That the hint is missing is clear, but what is the difference then?

**Comment by Robyt3** [comment]
> That the hint is missing is clear, but what is the difference then?

The default value for the hint appears to be false and according to the PR it should preserve the previous behaviour with `inp_ime_native_ui 0` in that case (https://github.com/libsdl-org/SDL/pull/4707). If 0 then SDL never calls `UILess_SetupSinks` (https://github.com/libsdl-org/SDL/blob/5fcd70578bdde3eaaa51aa4ab6c24ed08bd49db3/src/video/windows/SDL_windowskeyboard.c#L412-L416). I think the players tested with both values of `inp_ime_native_ui`, so we can assume that `UILess_SetupSinks` is not causing the crash.

@murpii It would be very helpful if some of those players could test with more different SDL versions to identify the newest SDL version that does not crash. We know that 2.0.16 works (which is used in 16.9) and 2.26.5 doesn't work (used in 17.0.2) so for releases between those versions it would be relevant to know if it crashes or not. Start with a fresh 17.0.2 version and then copy in the `SDL.dll` from `SDL2-2.X.Y-win32-x64.zip` from https://github.com/libsdl-org/SDL/releases for each version between 2.0.16 and 2.26.5 and test if it crashes with the `SDL.dll` from that version.

**Comment by Jupeyy** [comment]
Maybe my question was not clear. I want to know, if the behavior for the user feels exactly the same if:
- inp_ime_native_ui 1 with new SDL2
- using old SDL2

And i mean the bahavior outside of SDL2 code (be it buggy or not).

If I make a issue on SDL repo, I basically want to say:
We use the SDL2 versions exactly the same, they do exactly the same for the users. And still one version crashes, while the other works.

**Comment by murpii** [comment]
> @murpii It would be very helpful if some of those players could test with more different SDL versions to identify the newest SDL version that does not crash (and the oldest version that does crash). We know that 2.0.16 works (which is used in 16.9) and 2.26.5 doesn't work (used in 17.0.2) so for releases between those versions it would be relevant to know if they crash or not. Start with a fresh 17.0.2 release and then copy in the `SDL.dll` from `SDL2-2.X.Y-win32-x64.zip` from https://github.com/libsdl-org/SDL/releases for each version between 2.0.16 and 2.26.5 and test if it crashes with the `SDL.dll` from that version.

That's going to take some convincing, hm.

**Comment by murpii** [comment]
Findings:
The standalone version 17.0.3 (Non-steam) does not crash, even with SDL version 2.26.5
On steam, DDNet on version 17.0.3 with SDL version 2.0.20 works, all versions higher than that crash the game.

Testing was done using OpenGL 3.3.0

**Comment by Robyt3** [comment]
I could not reproduce the crash with the default Simplified/Traditional Chinese IMEs from Microsoft, Vulkan/OpenGL, `inp_ime_native_ui 0/1`, Steam/non-Steam.

The keyboard name in the screenshot is translated by Google Lens to "Microsoft Zhuyin" which seems to be Chinese name for "Microsoft Bopomofo", the default IME for Traditional Chinese that I also used. However, this IME doesn't seem to have a candidate window, so this should be unaffected by `inp_ime_native_ui`.

@murpii Does this crash happen consistently every time for the affected players? Are these players all using Traditional Chinese or is it also crashing for some with Simplified Chinese? Can some of them try if using Vulkan instead of OpenGL makes any difference?

**Comment by murpii** [comment]
>Does this crash happen consistently every time for the affected players?

No, it's very unconsistent. The crash happens in the middle of writing a message, randomly.

> Can some of them try if using Vulkan instead of OpenGL makes any difference?

Yes, it does make a difference. Using Vulkan the game no longer crashes (This applies to all people experiecing this specific issue.)
